pkgs/tools/security/pinentry/default.nix

```diff
@@ -1,13 +1,41 @@
 { fetchurl, fetchpatch, stdenv, lib, pkgconfig
-, libgpgerror, libassuan, libcap ? null, libsecret ? null, ncurses ? null,  gtk2 ? null, gcr ? null, qt ? null
-, enableEmacs ? false
+, libgpgerror, libassuan
+, ncurses, gtk2, qt
+, libcap ? null, libsecret ? null, gcr ? null
+, flavours ? [ "curses" "tty" "gtk2" "qt" "gnome3" "emacs" ]
 }:
 
+with stdenv.lib;
+
+assert isList flavours && flavours != [];
+
 let
-  mkFlag = pfxTrue: pfxFalse: cond: name: "--${if cond then pfxTrue else pfxFalse}-${name}";
+  mkFlag = pfxTrue: pfxFalse: cond: name:
+    "--${if cond then pfxTrue else pfxFalse}-${name}";
   mkEnable = mkFlag "enable" "disable";
   mkWith = mkFlag "with" "without";
+
+  mkEnablePinentry = f:
+    let
+      info = flavourInfo.${f};
+      inputs = info.buildInputs or [];
+      flag = flavourInfo.${f}.flag or null;
+      inputsSatifsfied = inputs == [] || all (f: !(isNull f)) inputs;
+    in
+      optionalString (flag != null)
+        (mkEnable (elem f flavours && inputsSatifsfied) ("pinentry-" + flag));
```

Comment:
```suggestion
        (mkEnable (elem f flavours && inputsSatisfied) ("pinentry-" + flag));
```

Reply:
done. we can simply assert `inputsSatisfied`, because `mkEnablePinentry` is only called for every `enabledFlavor(s)`